examples/llm_compression/torch/qat_with_lora/main_nls.py

```
@@ -0,0 +1,693 @@
+    :param batch_size: The batch size to be used during evaluation.
+    :return: A dictionary containing the evaluation results.
+    """
+    lm = HFLM(pretrained=model, tokenizer=tokenizer, batch_size=batch_size)
+    results = evaluator.simple_evaluate(lm, tasks=task, log_samples=False)["results"]
+    return results[task]
+
+
+def tokenize(
+    tokenizer: AutoTokenizer, prompt: str, add_eos_token: bool = True, max_length: int = 256
+) -> dict[str, list[int]]:
+    """
+    Tokenize the given prompt.
+
+    :param tokenizer: The tokenizer to use.
+    :param prompt: The prompt to tokenize.
+    :param add_eos_token: Whether to add an eos token.
+    :param max_length: The maximum length of the tokenized input.
+    :return: A dictionary containing tokenized input ids, attention mask, and labels.
+    """
+    result = tokenizer(
+        prompt,
+        truncation=True,
+        max_length=max_length,
+        padding=True,
+        return_tensors=None,
+    )
+    if result["input_ids"][-1] != tokenizer.eos_token_id and len(result["input_ids"]) < max_length and add_eos_token:
+        result["input_ids"].append(tokenizer.eos_token_id)
+        result["attention_mask"].append(1)
+
+    result["labels"] = result["input_ids"].copy()
+    return result
+
+
+def get_layer_id_vs_lora_quantizers_map(
+    model: nn.Module,
+) -> dict[int, list[Union["AsymmetricLoraQuantizer", "SymmetricLoraQuantizer"]]]:
+    """
+    Maps layer IDs to their corresponding LoRA quantizers.
+
+    :param model: The model containing LoRA quantizers.
+    :return: A dictionary mapping layer IDs to lists of LoRA quantizers.
+    """
+    hook_storage = get_hook_storage(model)
+    layer_id_vs_lora_quantizers_map = defaultdict(list)
+
+    for name, module in hook_storage.named_hooks():
+        if isinstance(module, (AsymmetricLoraQuantizer, SymmetricLoraQuantizer)) and (module.num_bits == 4):
+            layer_id = int(re.search(r"layers:(\d+):", name).group(1))
```

Please raise an error in case of not matched `layer_id` with the message that model is supposed to have a specific structure with Transformer blocks stored as follows: `self.layers = nn.ModuleList(...)`